frontend/packages/kubevirt-plugin/src/k8s/wrapper/common/wrapper.ts

```diff
@@ -1,26 +1,36 @@
 import * as _ from 'lodash';
+import { ensurePath } from '../utils/utils';
+import { omitEmpty } from '../../../utils/basic';
 
-export class Wrapper<RESOURCE extends {}> {
+export class Wrapper<RESOURCE extends {}, SELF extends Wrapper<RESOURCE, SELF>> {
   protected data: RESOURCE;
 
-  protected static defaultMergeWrappers = <A, B extends Wrapper<A>>(Clazz, wrappers: B[]): B => {
-    const nonEmptyWrappers = wrappers.filter((i) => i);
-    if (nonEmptyWrappers.length === 0) {
-      return new Clazz();
-    }
+  constructor(data?: RESOURCE | SELF, copy = false) {
+    const d = _.isFunction((data as any)?.asResource)
+      ? (data as SELF).asResource()
+      : (data as RESOURCE);
+    this.data = (d && copy ? _.cloneDeep(d) : d || {}) as any;
+  }
 
-    const mergedWrappers: A = _.merge({}, ...nonEmptyWrappers.map((i) => i.data));
+  public asResource = (copy = false): RESOURCE => (copy ? _.cloneDeep(this.data) : this.data);
```

Comment:
As discussed before, users of wrappers need to know about & use the right `copy` argument value.

An alternative (not sure how feasible) would be to use `Object.freeze` (recursive version, of course) to ensure the resource is immutable, if needed.

Reply:
yes, there are few alternatives how to approach this but each one of them comes with some drawbacks. At least this one is vanilla JS and is quite easy to use